src/main/java/com/microsoft/aad/adal4j/WSTrustRequest.java

```diff
@@ -77,13 +77,52 @@ static WSTrustResponse execute(String url, String username,
         return WSTrustResponse.parse(response, policy.getVersion());
     }
 
+    static WSTrustResponse execute(String mexURL,
+            String cloudAudienceUrn,
+            Proxy proxy,
+            SSLSocketFactory sslSocketFactory) throws Exception {
+
+        Map<String, String> headers = new HashMap<String, String>();
+        headers.put("Content-Type", "application/soap+xml");
+        headers.put("return-client-request-id", "true");
+
+        // Discover the policy for authentication using the Metadata Exchange Url.
+        String mexResponse = HttpHelper.executeHttpGet(log, mexURL, proxy, sslSocketFactory);
+
+        BindingPolicy policy = MexParser.getPolicyFromMexResponseForIntegrated(mexResponse);
```

Comment:
policy might not exist for the user, so instead of failing with NullPointerException , AuthenticationException with informative message should be returned.
if api is called for "managed"  - "Password is required for managed user"
if user discovery failed - "Unknown User Type"

Reply:
I don't have completed understandings of those error messages. I think it maybe a good idea for your guys to fix it, since the new `policy` variable follows exactly the same behavior as the existing `policy` variable in existing execute(...) function. If we need to fix the "policy" variable here, I guess you guys would also need to fix the `policy` variable in the existing execute(...) function.

Reply:
I don't have completed understandings of those error messages. I think it maybe a good idea for your guys to fix it, since the new `policy` variable follows exactly the same behavior as the existing `policy` variable in existing `execute(...)` function. If we need to fix the "policy" variable here, I guess you guys would also need to fix the `policy` variable in the existing `execute(...)` function.